plugins/hls-refactor-plugin/src/Development/IDE/Plugin/CodeAction/Extract.hs

```diff
@@ -0,0 +1,268 @@
+--  1. the range wraps an entire LHsExpr
+--  2. the range wraps an LHsExpr exactly
+--
+-- More explanation of (2)... If a range does exactly match any LHsExpr, then we know it is an invalid range for
+-- extract, since there is no whitespace padding in the range.
+removeWhitespacePadding :: Range -> T.Text -> UnpaddedRange
+removeWhitespacePadding range@(Range (Position sl sc) (Position el ec)) source =
+  let selectedTxt = textInRange range source
+      -- TODO We do not yet handle tabs gracefully
+      isWhiteSpace c = elem c [' ', '\n']
+      (paddingLeft, minimalOnLeft) = T.span isWhiteSpace selectedTxt
+      (paddingRight, _) = T.span isWhiteSpace $ T.reverse minimalOnLeft
+      whitespaceToDelta = \case
+        ' '  -> (0 :: Sum UInt, -1 :: Sum UInt)
+        '\n' -> (-1, 0)
+        _    -> error "impossible"
+      paddingToDelta = foldMap whitespaceToDelta . T.unpack
+      (getSum -> rowsL, getSum -> colsL) = paddingToDelta paddingLeft
+      (getSum -> rowsR, getSum -> colsR) = paddingToDelta paddingRight
+     in UnpaddedRange $ Range (Position (sl - rowsL) (sc - colsL)) (Position (el - rowsR) (ec - colsR))
+
+-- | Queries an AST for the topmost LHsExpr that exactly matches the given UnpaddedRange.
+tryFindLargestExprInRange :: UnpaddedRange -> GenericQ (Maybe (SrcSpan, LHsExpr GhcPs))
+tryFindLargestExprInRange range = everythingBut (<|>) $ mkQ (Nothing, False) firstContainedExprQ
+  where
+    -- When we find an LHsExpr that is fully contained by this range, we return the LHsExpr and early-exit the
+    -- syb traversal (by returning True)
+    firstContainedExprQ = \case
+      lexpr@(L (SrcSpanAnn _ span@(RealSrcSpan realSrcSpan _)) _) :: LHsExpr GhcPs
+        | rangeSrcSpan <- rangeToSrcSpan (fromString $ unpackFS $ srcSpanFile realSrcSpan) (unUnpaddedRange range)
+        , isSameSrcSpanModuloFile rangeSrcSpan span
+        -> (Just (span, lexpr), True)
+      _ -> (Nothing, False)
+
+-- | Ensures that there is at least one newline's difference between this LHsDecl and the previous decl. If the
+-- original decl was on the next line after the anchor, with no padding, we simply return with no modifications
+setAtLeastNewlineDp :: LHsDecl GhcPs -> LHsDecl GhcPs
+setAtLeastNewlineDp (L srcSpanAnn decl) = L (mapAnchor (maybe nextLineAnchor setAtLeastNewline) srcSpanAnn) decl
+  where
+    setAtLeastNewline (Anchor realSrcSpan ancOp) =
+              let ancOp' = case ancOp of
+                    UnchangedAnchor                 -> nextLine2Dp
+                    MovedAnchor (SameLine _)        -> nextLine2Dp
+                    MovedAnchor (DifferentLine 0 _) -> nextLine2Dp
+                    ancOp_                          -> ancOp_
+                in Anchor realSrcSpan ancOp'
+
+    nextLine2Dp = MovedAnchor (DifferentLine 2 0)
+    nextLineAnchor = generatedAnchor nextLine2Dp
+
+-- | Find all variable bindings in an AST.
```

Comment:
I see you asked in #ghc about this problem. I suggest making a GHC issue also explaining your difficulties. Then we can refer to it from here and it'll be useful in case this comes up again.